src/main/java/io/odpf/firehose/sink/objectstorage/writer/remote/ObjectStorageChecker.java

```diff
@@ -1,52 +1,131 @@
 package io.odpf.firehose.sink.objectstorage.writer.remote;
 
+import io.odpf.firehose.metrics.Instrumentation;
 import io.odpf.firehose.objectstorage.ObjectStorage;
+import io.odpf.firehose.objectstorage.ObjectStorageException;
+import io.odpf.firehose.objectstorage.gcs.exception.GCSException;
+import io.odpf.firehose.sink.objectstorage.writer.local.FileMeta;
+import io.odpf.firehose.sink.objectstorage.writer.local.Partition;
+import io.odpf.firehose.util.Clock;
 import lombok.AllArgsConstructor;
+import org.apache.commons.lang3.exception.ExceptionUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.io.IOException;
+import java.time.Instant;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Set;
 import java.util.concurrent.BlockingQueue;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
 import java.util.stream.Collectors;
 
+import static io.odpf.firehose.metrics.Metrics.*;
+import static io.odpf.firehose.sink.objectstorage.ObjectStorageMetrics.*;
+
 @AllArgsConstructor
 public class ObjectStorageChecker implements Runnable {
     private static final Logger LOGGER = LoggerFactory.getLogger(ObjectStorageChecker.class);
-    private final BlockingQueue<String> toBeFlushedToRemotePaths;
+
+    private final BlockingQueue<FileMeta> toBeFlushedToRemotePaths;
     private final BlockingQueue<String> flushedToRemotePaths;
     private final Set<ObjectStorageWriterWorkerFuture> remoteUploadFutures;
     private final ExecutorService remoteUploadScheduler;
     private final ObjectStorage objectStorage;
+    private final Clock clock;
+    private final Instrumentation instrumentation;
 
     @Override
     public void run() {
-        List<String> tobeFlushed = new ArrayList<>();
+        List<FileMeta> tobeFlushed = new ArrayList<>();
         toBeFlushedToRemotePaths.drainTo(tobeFlushed);
         remoteUploadFutures.addAll(tobeFlushed.stream()
-                .map(path -> new ObjectStorageWriterWorkerFuture(
-                        remoteUploadScheduler.submit(() -> objectStorage.store(path)), path)
+                .map(fileMeta -> {
+                            Runnable uploadTask = createUploadTask(fileMeta);
+                            Future<?> future = remoteUploadScheduler.submit(uploadTask);
+                            Instant startTime = clock.now();
```

Comment:
this instant will not calculate the correct upload time. The upload will be schedule and execute whenever the threads are free